File: output/GPT3.5-Turbo/xCodeEval_remove_4000/Java/Go/fc4bab2636b36a4bf15bb74a3a713a2c.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var mod = 998244353

func solve() {
	n := ni()
	K := ni()
	d := na(K)
	f := make([]int64, 20)
	for i := 0; i < K; i++ {
		for j := 0; j < K; j++ {
			f[10+d[i]-d[j]]++
		}
	}

	n /= 2
	g := make([]int64, 1)
	g[0] = 1
	for D := 17; D >= 0; D-- {
		g = clean(id6(g, g, mod, 3))
		if n<<~D < 0 {
			g = clean(id5(g, f))
		}
	}
	fmt.Println(n*10 < len(g)*int64(len(g)) && n*10 < int64(len(g)) ? g[n*10] : 0)
}

func clean(a []int64) []int64 {
	for i := len(a) - 1; i >= 0; i-- {
		if a[i] != 0 {
			if i == len(a)-1 {
				return a
			}
			return a[:i+1]
		}
	}
	return []int64{}
}

func id5(a []int64, b []int64) []int64 {
	c := make([]int64, len(a)+len(b)-1)
	big := 8 * int64(mod) * int64(mod)
	for i := 0; i < len(a); i++ {
		for j := 0; j < len(b); j++ {
			c[i+j] += a[i] * b[j]
			if c[i+j] >= big {
				c[i+j] -= big
			}
		}
	}
	for i := 0; i < len(c); i++ {
		c[i] %= int64(mod)
	}
	return c
}

var id8 = []int{1053818881, 1051721729, 1045430273, 1012924417, 1007681537, 1004535809, 998244353, 985661441, 976224257, 975175681}
var id4 = []int{7, 6, 3, 5, 3, 3, 3, 3, 3, 17}

func id6(a []int64, b []int64, P int, g int) []int64 {
	m := max(2, 1<<(32-__builtin_clz(max(len(a), len(b))-1)<<2))
	fa := id9(a, m, false, P, g)
	fb := a
	if a != b {
		fb = id9(b, m, false, P, g)
	}
	for i := 0; i < m; i++ {
		fa[i] = fa[i] * fb[i] % int64(P)
	}
	return id9(fa, m, true, P, g)
}

func id3(a []int64, b []int64) []int64 {
	USE := 2
	m := max(2, 1<<(32-__builtin_clz(max(len(a), len(b))-1)<<2))
	fs := make([][]int64, USE)
	for k := 0; k < USE; k++ {
		P := id8[k]
		g := id4[k]
		fa := id9(a, m, false, P, g)
		fb := a
		if a != b {
			fb = id9(b, m, false, P, g)
		}
		for i := 0; i < m; i++ {
			fa[i] = fa[i] * fb[i] % int64(P)
		}
		fs[k] = id9(fa, m, true, P, g)
	}

	mods := id8[:USE]
	gammas := id1(mods)
	buf := make([]int, USE)
	for i := 0; i < len(fs[0]); i++ {
		for j := 0; j < USE; j++ {
			buf[j] = int(fs[j][i])
		}
		res := id7(buf, mods, gammas)
		ret := int64(0)
		for j := len(res) - 1; j >= 0; j-- {
			ret = ret*int64(mods[j]) + res[j]
		}
		fs[0][i] = ret
	}
	return fs[0]
}

func id3(a []int64, b []int64, USE int, mod int) []int64 {
	m := max(2, 1<<(32-__builtin_clz(max(len(a), len(b))-1)<<2))
	fs := make([][]int64, USE)
	for k := 0; k < USE; k++ {
		P := id8[k]
		g := id4[k]
		fa := id9(a, m, false, P, g)
		fb := a
		if a != b {
			fb = id9(b, m, false, P, g)
		}
		for i := 0; i < m; i++ {
			fa[i] = fa[i] * fb[i] % int64(P)
		}
		fs[k] = id9(fa, m, true, P, g)
	}

	mods := id8[:USE]
	gammas := id1(mods)
	buf := make([]int, USE)
	for i := 0; i < len(fs[0]); i++ {
		for j := 0; j < USE; j++ {
			buf[j] = int(fs[j][i])
		}
		res := id7(buf, mods, gammas)
		ret := int64(0)
		for j := len(res) - 1; j >= 0; j-- {
			ret = (ret*int64(mods[j]) + res[j]) % int64(mod)
		}
		fs[0][i] = ret
	}
	return fs[0]
}

func id9(src []int64, n int, inverse bool, P int, g int) []int64 {
	dst := make([]int64, n)

	h := __builtin_ctz(n)
	K := int64(1) << 32
	M := K * K / int64(P)

	wws := make([]int64, 1<<(h-1))
	dw := int64(0)
	if inverse {
		dw = pow(int64(g), int64(P-1-(P-1)/n), int64(P))
	} else {
		dw = pow(int64(g), int64((P-1)/n), int64(P))
	}
	w := (1 << 32) % P
	for k := 0; k < 1<<(h-1); k++ {
		wws[k] = int64(w)
		w = modh(w*dw, M, h, P)
	}

	J := invl(int64(P), int64(1)<<32)
	for i := 0; i < h; i++ {
		for j := 0; j < 1<<i; j++ {
			for k, s, t := 0, j<<h-i, j<<h-i|1<<(h-i-1); k < 1<<(h-i-1); k, s, t = k+1, s+1, t+1 {
				u := (dst[s] - dst[t] + 2*int64(P)) * wws[k]
				dst[s] += dst[t]
				if dst[s] >= 2*int64(P) {
					dst[s] -= 2 * int64(P)
				}

				Q := (u << 32) * J >> 32
				dst[t] = (u >> 32) - (Q*int64(P) >> 32) + int64(P)
			}
		}
		if i < h-1 {
			for k := 0; k < 1<<(h-i-2); k++ {
				wws[k] = wws[k*2]
			}
		}
	}

	for i := 0; i < n; i++ {
		if dst[i] >= int64(P) {
			dst[i] -= int64(P)
		}
	}
	for i := 0; i < n; i++ {
		rev := __builtin_bswap32(uint32(i)) >> (32 - h)
		if i < int(rev) {
			d := dst[i]
			dst[i] = dst[rev]
			dst[rev] = d
		}
	}

	if inverse {
		in := invl(int64(n), int64(P))
		for i := 0; i < n; i++ {
			dst[i] = modh(dst[i]*in, M, h, P)
		}
	}

	return dst
}

func id1(m []int) []int64 {
	n := len(m)
	gamma := make([]int64, n)
	for k := 1; k < n; k++ {
		prod := int64(1)
		for i := 0; i < k; i++ {
			prod = prod * int64(m[i]) % int64(m[k])
		}
		gamma[k] = invl(prod, int64(m[k]))
	}
	return gamma
}

func id7(u []int, m []int, gamma []int64) []int64 {
	n := len(u)
	v := make([]int64, n)
	v[0] = int64(u[0])
	for k := 1; k < n; k++ {
		temp := v[k-1]
		for j := k - 2; j >= 0; j-- {
			temp = (temp*int64(m[j]) + v[j]) % int64(m[k])
		}
		v[k] = (int64(u[k]) - temp) * gamma[k] % int64(m[k])
		if v[k] < 0 {
			v[k] += int64(m[k])
		}
	}
	return v
}

func pow(a, n, mod int64) int64 {
	ret := int64(1)
	x := 63 - __builtin_clz(int(n))
	for ; x >= 0; x-- {
		ret = ret * ret % mod
		if n<<63-x < 0 {
			ret = ret * a % mod
		}
	}
	return ret
}

func invl(a, mod int64) int64 {
	b := mod
	p := int64(1)
	q := int64(0)
	for b > 0 {
		c := a / b
		d := a
		a = b
		b = d % b
		d = p
		p = q
		q = d - c*q
	}
	if p < 0 {
		p += mod
	}
	return p
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func modh(a, M int64, h, mod int) int64 {
	r := a - ((M*(a&((1<<31)-1))>>31)+M*(a>>31)>>(h-31))*int64(mod)
	if r < int64(mod) {
		return r
	}
	return r - int64(mod)
}

func main() {
	solve()
}

var reader = bufio.NewReader(os.Stdin)
var scanner = bufio.NewScanner(os.Stdin)

func init() {
	scanner.Split(bufio.ScanWords)
}

func scanString() string {
	scanner.Scan()
	return scanner.Text()
}

func scanInt() int {
	scanner.Scan()
	num, _ := strconv.Atoi(scanner.Text())
	return num
}

func scanInt64() int64 {
	scanner.Scan()
	num, _ := strconv.ParseInt(scanner.Text(), 10, 64)
	return num
}

func scanFloat() float64 {
	scanner.Scan()
	num, _ := strconv.ParseFloat(scanner.Text(), 64)
	return num
}

func scanBytes() []byte {
	reader.Read(inbuf)
	return inbuf
}

func ni() int {
	scanner.Scan()
	num, _ := strconv.Atoi(scanner.Text())
	return num
}

func nl() int64 {
	scanner.Scan()
	num, _ := strconv.ParseInt(scanner.Text(), 10, 64)
	return num
}

func nf() float64 {
	scanner.Scan()
	num, _ := strconv.ParseFloat(scanner.Text(), 64)
	return num
}

func ns() string {
	scanner.Scan()
	return scanner.Text()
}

func na(n int) []int {
	a := make([]int, n)
	for i := 0; i < n; i++ {
		a[i] = ni()
	}
	return a
}

func nlArr(n int) []int64 {
	a := make([]int64, n)
	for i := 0; i < n; i++ {
		a[i] = nl()
	}
	return a
}

func nfArr(n int) []float64 {
	a := make([]float64, n)
	for i := 0; i < n; i++ {
		a[i] = nf()
	}
	return a
}

func nsArr(n int) []string {
	a := make([]string, n)
	for i := 0; i < n; i++ {
		a[i] = ns()
	}
	return a
}

func printString(s string) {
	fmt.Print(s)
}

func printInt(n int) {
	fmt.Print(n)
}

func printInt64(n int64) {
	fmt.Print(n)
}

func printFloat(f float64) {
	fmt.Print(f)
}

func printBytes(b []byte) {
	os.Stdout.Write(b)
}

func printlnString(s string) {
	fmt.Println(s)
}

func printlnInt(n int) {
	fmt.Println(n)
}

func printlnInt64(n int64) {
	fmt.Println(n)
}

func printlnFloat(f float64) {
	fmt.Println(f)
}

func printlnBytes(b []byte) {
	os.Stdout.Write(b)
}

func printfString(format string, a ...interface{}) {
	fmt.Printf(format, a...)
}

func printfInt(format string, a ...interface{}) {
	fmt.Printf(format, a...)
}

func printfInt64(format string, a ...interface{}) {
	fmt.Printf(format, a...)
}

func printfFloat(format string, a ...interface{}) {
	fmt.Printf(format, a...)
}

func printfBytes(format string, a ...interface{}) {
	fmt.Printf(format, a...)
}

func println() {
	fmt.Println()
}

func pow(a, n, mod int64) int64 {
	ret := int64(1)
	x := 63 - __builtin_clz(int(n))
	for ; x >= 0; x-- {
		ret = ret * ret % mod
		if n<<63-x < 0 {
			ret = ret * a % mod
		}
	}
	return ret
}

func invl(a, mod int64) int64 {
	b := mod
	p := int64(1)
	q := int64(0)
	for b > 0 {
		c := a / b
		d := a
		a = b
		b = d % b
		d = p
		p = q
		q = d - c*q
	}
	if p < 0 {
		p += mod
	}
	return p
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func __builtin_clz(x int) int {
	return 31 - bits.Len32(uint32(x))
}

func __builtin_bswap32(x uint32) uint32 {
	return bits.Reverse32(x)
}

func __builtin_ctz(x int) int {
	return bits.TrailingZeros32(uint32(x))
}

func __builtin_popcount(x int) int {
	return bits.OnesCount32(uint32(x))
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/Java/Python/fc4bab2636b36a4bf15bb74a3a713a2c.py
import sys
from typing import List

def solve(n: int, K: int, d: List[int]) -> int:
    f = [0] * 20
    for i in range(K):
        for j in range(K):
            f[10 + d[i] - d[j]] += 1
    
    n //= 2
    g = [1]
    for D in range(17, -1, -1):
        g = clean(id6(g, g, mod, 3))
        if n << ~D < 0:
            g = clean(id5(g, f))
    
    return g[n * 10] if n * 10 < len(g) else 0

def clean(a: List[int]) -> List[int]:
    for i in range(len(a) - 1, -1, -1):
        if a[i] != 0:
            if i == len(a) - 1:
                return a
            return a[:i+1]
    return []

def id5(a: List[int], b: List[int]) -> List[int]:
    c = [0] * (len(a) + len(b) - 1)
    big = 8 * mod * mod
    for i in range(len(a)):
        for j in range(len(b)):
            c[i + j] += a[i] * b[j]
            if c[i + j] >= big:
                c[i + j] -= big
    for i in range(len(c)):
        c[i] %= mod
    return c

def id6(a: List[int], b: List[int], P: int, g: int) -> List[int]:
    m = max(2, (1 << (len(bin(max(len(a), len(b))) - 2)))) << 2
    fa = id9(a, m, False, P, g)
    fb = fa if a == b else id9(b, m, False, P, g)
    for i in range(m):
        fa[i] = fa[i] * fb[i] % P
    return id9(fa, m, True, P, g)

def id9(src: List[int], n: int, inverse: bool, P: int, g: int) -> List[int]:
    dst = src.copy()
    h = n.bit_length() - 1
    K = (1 << (P.bit_length() - 1)) << 1
    H = K.bit_length() * 2
    M = K * K // P

    wws = [0] * (1 << (h - 1))
    dw = pow(g, P - 1 - (P - 1) // n, P) if inverse else pow(g, (P - 1) // n, P)
    w = (1 << 32) % P
    for k in range(1 << (h - 1)):
        wws[k] = w
        w = modh(w * dw, M, H, P)
    
    J = invl(P, 1 << 32)
    for i in range(h):
        for j in range(1 << i):
            for k in range(1 << (h - i - 1)):
                s = j << (h - i)
                t = s | (1 << (h - i - 1))
                u = (dst[s] - dst[t] + 2 * P) * wws[k]
                dst[s] += dst[t]
                if dst[s] >= 2 * P:
                    dst[s] -= 2 * P
                Q = (u << 32) * J >> 32
                dst[t] = (u >> 32) - (Q * P >> 32) + P
        
        if i < h - 1:
            for k in range(1 << (h - i - 2)):
                wws[k] = wws[k * 2]
        
    for i in range(n):
        if dst[i] >= P:
            dst[i] -= P
    for i in range(n):
        rev = int(bin(i)[2:].zfill(h)[::-1], 2)
        if i < rev:
            dst[i], dst[rev] = dst[rev], dst[i]
    
    if inverse:
        inv = invl(n, P)
        for i in range(n):
            dst[i] = modh(dst[i] * inv, M, H, P)
    
    return dst

def modh(a: int, M: int, h: int, mod: int) -> int:
    r = a - ((M * (a & ((1 << 31) - 1)) >> 31) + M * (a >> 31) >> (h - 31)) * mod
    return r if r < mod else r - mod

def id1(m: List[int]) -> List[int]:
    n = len(m)
    gamma = [0] * n
    for k in range(1, n):
        prod = 1
        for i in range(k):
            prod = prod * m[i] % m[k]
        gamma[k] = invl(prod, m[k])
    return gamma

def id7(u: List[int], m: List[int], gamma: List[int]) -> List[int]:
    n = len(u)
    v = [0] * n
    v[0] = u[0]
    for k in range(1, n):
        temp = v[k - 1]
        for j in range(k - 2, -1, -1):
            temp = (temp * m[j] + v[j]) % m[k]
        v[k] = (u[k] - temp) * gamma[k] % m[k]
        if v[k] < 0:
            v[k] += m[k]
    return v

def pow(a: int, n: int, mod: int) -> int:
    ret = 1
    x = 63 - n.bit_length()
    while x >= 0:
        ret = ret * ret % mod
        if n << (63 - x) < 0:
            ret = ret * a % mod
        x -= 1
    return ret

def invl(a: int, mod: int) -> int:
    b = mod
    p, q = 1, 0
    while b > 0:
        c = a // b
        d = a
        a = b
        b = d % b
        d = p
        p = q
        q = d - c * q
    return p + mod if p < 0 else p

mod = 998244353

def main():
    n, K = map(int, input().split())
    d = list(map(int, input().split()))
    result = solve(n, K, d)
    print(result)

if __name__ == "__main__":
    main()
